adds LinkedProfile and LinkedSite boilerplate

toyhammered opened the pull request:
I am not entirely sure _what_ to be testing in the models aside for their relationships.
🙏 
@NuckChorris

toyhammered commented on app/resources/linked_site_resource.rb:
```diff
@@ -0,0 +1,4 @@
+class LinkedSiteResource < BaseResource
+  immutable
```
Was not sure if this was right.

NuckChorris replied:
It is

toyhammered commented on app/resources/linked_profile_resource.rb:
```diff
@@ -0,0 +1,20 @@
+  has_one :user
+  has_one :linked_site
+
+  filter :self, apply: -> (_r, _v, options) {
```
don't forget to look into if it needs to be `_r` or `records`

NuckChorris replied:
Turns out, it's supposed to be `records`, nice catch!

NuckChorris commented on app/models/linked_profile.rb:
```diff
@@ -0,0 +1,30 @@
+#  fk_rails_25de88e967  (linked_site_id => linked_sites.id)
+#
+
+class LinkedProfile < ActiveRecord::Base
```
Could you change these to inherit from `ApplicationRecord`?

NuckChorris commented on app/models/linked_profile.rb:
```diff
@@ -0,0 +1,30 @@
+
+class LinkedProfile < ActiveRecord::Base
+  belongs_to :user
+  belongs_to :linked_site
```
These should have `required: true` since that's becoming the default in Rails 5

NuckChorris commented on app/models/linked_site.rb:
```diff
@@ -0,0 +1,18 @@
+#
+
+class LinkedSite < ActiveRecord::Base
+  LINK_TYPE = %w[username OAuth2].freeze
```
`oauth2` — symbols like this shouldn't be uppercase

Also should be a `%i` literal for symbols, not `%w`

And it's not necessary to declare it as a constant, you can also just do `enum link_type: %i[username oauth2]`

NuckChorris commented on app/models/linked_profile.rb:
```diff
@@ -0,0 +1,30 @@
+# == Schema Information
+#
+# Table name: linked_profiles
```
There should probably be a `public` boolean column too

toyhammered replied:
is `public` reserved cause it is causing me issues when I am trying to test it.

NuckChorris commented on app/policies/linked_profile_policy.rb:
```diff
@@ -0,0 +1,11 @@
+class LinkedProfilePolicy < ApplicationPolicy
+  def show?
+    record.user == user || user.has_role?(:admin)
```
Should probably be `record.public? || record.user == user || is_admin?` because:
1. linked profiles need to be public so we can show 'em on your profile
2. `is_admin?` is a shorthand for `user.has_role?(:admin)`

toyhammered replied:
isn't `is_admin?` only accessible in `create, update, destroy` because of the `alias_method`?

```
  def is_admin? # rubocop:disable Style/PredicateName
    user && user.has_role?(:admin, model_class)
  end
  alias_method :create?, :is_admin?
  alias_method :update?, :is_admin?
  alias_method :destroy?, :is_admin?
```

NuckChorris replied:
No, `alias_method from, to` means "create a method with `from` name that just defers to `to`"

So that basically creates the following method:

``` ruby
def create?(*args)
  return is_admin?(*args)
end
```

It does not affect scoping, it's not magic.  It literally creates an _alias_ for a _method_ ;)

toyhammered replied:
aye aye captain! 

(now a bit off-topic) why do you do this then?

```
class LibraryEntryPolicy < ApplicationPolicy
  def show?
    # Yes, if you own it or are admin
    return true if record.user == user || user.has_role?(:admin)
    # No, if it's private
    return false if record.private?
    # No, if it's nsfw and you're not a perv
    return false if user.sfw_filter? && record.media.nsfw?
    # Otherwise, yeah
    true
  end
...
```

NuckChorris replied:
Sometimes I make mistakes

toyhammered replied:
![impossibru](http://img.imgur.com/yjZ18Ux.jpg)

NuckChorris commented on app/resources/linked_profile_resource.rb:
```diff
@@ -0,0 +1,20 @@
+    if current_user == _model
+      super
+    else
+      super - :token
```
You missed an important detail here: `:token` needs to be `[:token]` — you can't subtract a symbol from an array, but you can subtract an array from an array

NuckChorris commented on db/migrate/20161019004726_create_linked_profiles.rb:
```diff
@@ -0,0 +1,16 @@
+  def change
+    create_table :linked_profiles do |t|
+      t.references :user, index: true, foreign_key: true
+      t.references :linked_site, index: true, foreign_key: true
```
These two should probably have `null: false`

NuckChorris commented on db/migrate/20161019004726_create_linked_profiles.rb:
```diff
@@ -0,0 +1,16 @@
+      t.references :user, index: true, foreign_key: true
+      t.references :linked_site, index: true, foreign_key: true
+
+      t.integer :external_user_id
```
This should be a string... I know, confusing, but some other sites are going to be dipshits and not provide an integer ID.  For example, some sites might have UUIDs like `59208787-86ec-4aef-97ae-cede8e72ecc4` or weird base64 crap like `Mmvofg3gGq0c4w`

We need to be ready to accept this

NuckChorris commented on app/models/linked_profile.rb:
```diff
@@ -0,0 +1,30 @@
+class LinkedProfile < ActiveRecord::Base
+  belongs_to :user
+  belongs_to :linked_site
+end
```
I would recommend tossing in a validation of `validates :url, presence: true, if: :public?` (once you add a `public` column)

NuckChorris commented on db/migrate/20161019004726_create_linked_profiles.rb:
```diff
@@ -0,0 +1,16 @@
+      t.integer :external_user_id
+      t.string :url
+      t.boolean :share_to
+      t.boolean :share_from
```
These should be `null: false` w/ a default probably, like `default: false`

toyhammered replied:
I am assuming this only references `lines 9 & 10`?

NuckChorris commented on db/migrate/20161019003515_create_linked_sites.rb:
```diff
@@ -0,0 +1,12 @@
+      t.string :name
+      t.boolean :share_to
+      t.boolean :share_from
+      t.integer :link_type
```
All of these should have `null: false` on them, since they're all switches.

Remember: a switch is either on or off.  You don't want a switch that can randomly cease existence

toyhammered replied:
I am assuming this only references `lines 5 - 7`

NuckChorris replied:
No, I mean all of them.  What good is a website without a name?

toyhammered replied:
:|

NuckChorris reviewed (changes requested):
A good start :+1:

toyhammered committed:
fixes all of nucks previous comments

toyhammered committed:
fixes merge conflicts

toyhammered commented on spec/controllers/users_controller_spec.rb:
```diff
@@ -1,3 +1,79 @@
+#  name                        :string(255)
+#  ninja_banned                :boolean          default(FALSE)
+#  onboarded                   :boolean          default(FALSE), not null
+#  past_names                  :string           default([]), not null, is an Array
```
go fuck yourself hound...

toyhammered commented:
I have a feeling my mistake is not doing `validates :name, presence: true` and for all the other `null: false` fields added to the database (is this why FactoryGirl is failing?)

NuckChorris commented on app/models/user.rb:
```diff
@@ -89,6 +87,7 @@ class User < ApplicationRecord
     dependent: :destroy
   has_many :following, class_name: 'Follow', foreign_key: 'follower_id',
     dependent: :destroy
+  has_many :linked_profiles
```
Pretty much all `has_many` associations ought to have `dependent: :destroy`.  I forgot about this until recently, but it's important, because otherwise things will break if we try to delete a user

NuckChorris commented on app/resources/linked_profile_resource.rb:
```diff
@@ -0,0 +1,20 @@
+  filter :self, apply: -> (records, _v, options) {
+    current_user = options[:context][:user]
+    records.where(id: current_user&.id) || User.none
+  }
```
You don't need a `self` filter for this, the client has the user_id and can use it for querying (or can just request `include=linkedProfile` on a user)

NuckChorris commented on spec/models/linked_profile_spec.rb:
```diff
@@ -0,0 +1,46 @@
+
+  describe 'validates url' do
+    context 'if public' do
+      before { allow(subject).to receive(:public?).and_return(true) }
```
You can better express this by doing this:

`subject { described_class.new(public: true) }`

Nested blocks can have different subjects from their parents

NuckChorris commented on spec/factories/linked_sites.rb:
```diff
@@ -0,0 +1,17 @@
+#
+
+FactoryGirl.define do
+  factory :linked_site do
```
You should generate valid fake data for columns without defaults here, like:

``` ruby
name { Faker::Company.name }
link_type { LinkedSite.link_types.sample }
```

NuckChorris commented on spec/factories/linked_profiles.rb:
```diff
@@ -0,0 +1,33 @@
+FactoryGirl.define do
+  factory :linked_profile do
+    user
+    linked_site
```
You should generate fake data here, like this:

``` ruby
url { Faker::Internet.url }
external_user_id { Faker::Internet.user_name }
```

NuckChorris commented on app/models/linked_site.rb:
```diff
@@ -0,0 +1,16 @@
+#
+
+class LinkedSite < ApplicationRecord
+  enum link_type: %i[username oauth2]
```
Even if it's not going to be touched by most users, this should definitely have at least basic presence validations for not-null columns

NuckChorris commented on app/models/linked_profile.rb:
```diff
@@ -0,0 +1,33 @@
+  belongs_to :user, required: true
+  belongs_to :linked_site, required: true
+
+  validates :url, presence: true, if: :public?
```
You've also marked `external_user_id` not-null, so there should be a presence validation for that too.

Also, if `url` isn't always required then it shouldn't be marked `not null` — that'll cause :boom:

toyhammered replied:
Isn't it required though if public is true (this shouldn't blow up because of the failsafe if statement?)

NuckChorris commented on spec/models/linked_site_spec.rb:
```diff
@@ -0,0 +1,18 @@
+require 'rails_helper'
+
+RSpec.describe LinkedSite, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
```
Please add some examples or delete this :P

toyhammered commented on app/models/linked_profile.rb:
```diff
@@ -0,0 +1,33 @@
+  belongs_to :user, required: true
+  belongs_to :linked_site, required: true
+
+  validates :url, presence: true, if: :public?
```
is url not going to be required if public is false?

toyhammered committed:
Attempts to fix minor rubocop errors + database
  - Schema is incorrect, needs to remove null: false from url
  - migration has been updated to reflect change
  - factories updated with fake data if null: false and no default

toyhammered committed:
Fixes rubocop errors, database errors
  - removes not null constraint on url
  - adds rubocop metrics thing

NuckChorris commented on app/policies/linked_profile_policy.rb:
```diff
@@ -0,0 +1,11 @@
+class LinkedProfilePolicy < ApplicationPolicy
+  def show?
+    record.user == user || record.public? || is_admin?
+  end
```
With Pundit-Resources we no longer use `#show?`

Instead, add a `Scope` class like you see in the `LibraryEntryPolicy`.  The logic should be pretty similar.

NuckChorris commented on app/models/installment.rb:
```diff
@@ -14,7 +13,6 @@
 #  index_installments_on_franchise_id             (franchise_id)
 #  index_installments_on_media_type_and_media_id  (media_type,media_id)
 #
-# rubocop:enable Metrics/LineLength
```
Could you undo all the places you removed `# rubocop:enable Metrics/LineLength`?

NuckChorris commented on spec/controllers/anime_controller_spec.rb:
```diff
@@ -1,3 +1,47 @@
+#  index_anime_on_wilson_ci   (average_rating)
+#
+# rubocop:enable Metrics/LineLength
+
```
Could you remove the annotations from `spec/controllers/*`

P.S. `git add -p` is your friend :wink:

NuckChorris reviewed (changes requested):
Pretty much all the logic is right, just got a little bit left to go on the cleanup :+1:

toyhammered committed:
Adds/Removes Rubocop metrics length where needed.
  - Starting on changing show? to scope

toyhammered committed:
Changes Public to Private in LinkedProfile Model

toyhammered committed:
Adds Pundit scope

NuckChorris commented on app/policies/linked_profile_policy.rb:
```diff
@@ -0,0 +1,25 @@
+
+  class Scope < Scope
+    def resolve
+      return scope.where(private: true) unless user
```
This seems backwards.  Limit anonymous users (`unless user`) to a scope where `private: true`?  So they can _only_ see private stuff?

toyhammered replied:
oh.. I am an idiot was thinking backwards

NuckChorris commented on spec/factories/installments.rb:
```diff
@@ -15,7 +14,6 @@
 #  index_installments_on_franchise_id             (franchise_id)
 #  index_installments_on_media_type_and_media_id  (media_type,media_id)
 #
-# rubocop:enable Metrics/LineLength
```
Missed a spot 😛

toyhammered replied:
🖕

toyhammered committed:
Fixes LinkedProfile Scope
  - changes tests to use private
  - adds rubocop metrics where needed

NuckChorris commented:
LGTM :tada:
